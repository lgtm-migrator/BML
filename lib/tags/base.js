const { BMLStream } = require('../BMLStream');
const { Context } = require('../context');
const { NotImplementedError } = require('../errors');
const { ensureTag } = require('../utils');

function prepareContext(contextData) {
  let context = new Context();
  if (contextData) {
    if (contextData instanceof Context) {
      context = contextData;
    } else {
      context.fill(contextData);
    }
  }
  return context;
}

class Tag {
  parse(data, contextData = {}) {
    const stream = new BMLStream(data);
    const context = prepareContext(contextData);
    return this._parse_(stream, context);
  }

  _parse_(_stream, _context) {
    throw new NotImplementedError(
      `_parse_ is not implemented in ${this.constructor.name}`
    );
  }

  pack(data, contextData = {}) {
    const stream = new BMLStream();
    const context = prepareContext(contextData);
    this._pack_(stream, data, context);
    return stream.finalize();
  }

  _pack_(_stream, _data, _context) {
    throw new NotImplementedError(
      `_parse_ is not implemented in ${this.constructor.name}`
    );
  }

  named(name) {
    this.name = name;
    return this;
  }

  toString() {
    return `${this.constructor.name} {
      ${this.name}
    }`;
  }
}

class CompositeTag extends Tag {
  constructor(...subTags) {
    super();
    this.subTags = subTags.map(ensureTag);
  }
}

class Adapter extends Tag {
  constructor(subTag) {
    super();
    this.tag = ensureTag(subTag);
  }

  _parse_(stream, context) {
    const data = this.tag._parse_(stream, context);
    return this._decode_(data, context);
  }

  _pack_(stream, data, context) {
    const encoded = this._encode_(data, context);
    this.tag._pack_(stream, encoded, context);
  }

  _decode_(_data, _context) {
    throw new NotImplementedError(
      `_decode_ is not implemented in ${this.constructor.name}`
    );
  }

  _encode_(_data, _context) {
    throw new NotImplementedError(
      `_encode_ is not implemented in ${this.constructor.name}`
    );
  }
}

function markTagFn(fn) {
  Object.defineProperty(fn, 'isTagFn', { get: () => true });
}

function tag(TagClass, args = []) {
  const tagFn = function tagFn(input) {
    let name;
    if (typeof input === 'string') {
      name = input; // calling tag with name as usual function
    } else if (Array.isArray(input) && input.raw) {
      [name] = input; // calling tag as template fn
    }
    function builderFn() {
      return new TagClass(...args).named(name);
    }
    markTagFn(builderFn);
    if (input === undefined) {
      return builderFn();
    }
    return builderFn;
  };
  Object.defineProperty(tagFn, 'name', {
    value: TagClass.name.toLowerCase(),
  });
  markTagFn(tagFn);
  return tagFn;
}

function tagWithParams(TagClass) {
  const tagParamFn = function tagParamFn(...args) {
    return tag(TagClass, args);
  };
  Object.defineProperty(tagParamFn, 'isTagFn', {
    get: () => {
      throw new Error(`params not passed for ${TagClass.name}`);
    },
  });
  return tagParamFn;
}

module.exports = {
  Tag,
  CompositeTag,
  Adapter,
  tag,
  tagWithParams,
};
