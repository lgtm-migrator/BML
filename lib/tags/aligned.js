const { Tag, tagWithParams } = require('./base');
const { ensureTag, functor } = require('../utils');

class Aligned extends Tag {
  constructor(subTag, modulus) {
    super();
    this.subTag = ensureTag(subTag);
    this.modulus = functor(modulus);
  }

  _parse_(stream, context) {
    const start_position = stream.tell();
    const value = this.subTag._parse_(stream, context);
    const end_position = stream.tell();
    const modulus = this.modulus(context);
    const pad = (end_position - start_position) % modulus;
    stream.seek(end_position + pad);
    return value;
  }

  _pack_(stream, data, context) {
    const start_position = stream.tell();
    this.subTag._pack_(stream, data, context);
    const end_position = stream.tell();
    const modulus = this.modulus(context);
    const pad = (end_position - start_position) % modulus;
    stream.seek(end_position + pad);
  }
}

module.exports = {
  aligned: tagWithParams(Aligned),
};
