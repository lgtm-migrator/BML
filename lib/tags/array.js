const { Tag, tagWithParams } = require('./base');
const { Context } = require('../Context');
const { functor, range, ensureTag } = require('../utils');

// Homogenous array of elements, similar to C# generic T[].
class ArrayTag extends Tag {
  constructor(tag, count) {
    super();
    this.tag = ensureTag(tag);
    this.count = functor(count);
  }

  _parse_(stream, context) {
    const count = this.count(context);
    const listContext = new Context(context);
    const data = [];
    for (const index of range(count)) {
      listContext.set('index', index);
      const subData = this.tag._parse_(stream, listContext);
      data.push(subData);
    }
    return data;
  }

  _pack_(stream, data, context) {
    const count = this.count(context);
    const listContext = new Context(context);
    for (const index of range(count)) {
      listContext.set('index', index);
      const subData = data[index];
      this.tag._pack_(stream, subData, listContext);
    }
  }

  _size_(context) {
    const listContext = new Context(context);
    return this.count(context) * this.tag._size_(listContext);
  }
}

module.exports = {
  array: tagWithParams(ArrayTag),
};
