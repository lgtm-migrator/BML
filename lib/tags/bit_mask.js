const { Tag, tagWithParams } = require('./base');
const { ensureTag } = require('../utils');

class BitMask extends Tag {
  constructor(subTag, bitMap) {
    super();
    this.subTag = ensureTag(subTag);
    this.bitMap = bitMap;
  }

  _parse_(stream, context) {
    const parsedValue = this.subTag._parse_(stream, context);
    const result = {};
    for (const key of Object.keys(this.bitMap)) {
      result[key] = Boolean(parsedValue & this.bitMap[key]);
    }
    return result;
  }

  _pack_(stream, data, context) {
    let value = 0;
    for (const key of Object.keys(this.bitMap)) {
      if (data[key]) {
        value |= this.bitMap[key];
      }
    }
    this.subTag._pack_(stream, value, context);
  }
}

module.exports = {
  bit_mask: tagWithParams(BitMask),
};
