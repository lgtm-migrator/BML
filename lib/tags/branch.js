const { Tag, tagWithParams } = require('./base');
const { functor, ensureTag } = require('../utils');

class Branch extends Tag {
  constructor(selector, cases, defaultTag) {
    super();
    this.selector = functor(selector);
    this.cases = cases;
    this.defaultTag = defaultTag;
  }

  _parse_(stream, context) {
    const caseKey = this.selector(context);
    let useCase = this.cases[caseKey];
    if (!useCase && this.defaultTag) {
      useCase = this.defaultTag;
    }
    if (useCase) {
      useCase = ensureTag(useCase);
      return useCase._parse_(stream, context);
    }
    return null;
  }

  _pack_(stream, data, context) {
    const caseKey = this.selector(context);
    let useCase = this.cases[caseKey];
    if (useCase) {
      useCase = ensureTag(useCase);
      useCase._pack_(stream, data, context);
    }
  }
}

module.exports = {
  branch: tagWithParams(Branch),
};
