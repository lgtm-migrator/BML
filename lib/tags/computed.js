const { Tag, tagWithParams } = require('./base');
const { functor } = require('../utils');

class Computed extends Tag {
  constructor(compute) {
    super();
    this.compute = functor(compute);
  }

  _parse_(_stream, context) {
    const computedValue = this.compute(context);
    if (this.name) {
      context.set(this.name, computedValue);
    }
    return computedValue;
  }

  _pack_(stream, _data, context) {
    this._parse_(stream, context);
  }

  _size_(_context) {
    return 0;
  }
}

const computed = tagWithParams(Computed);

module.exports = {
  computed,
};
