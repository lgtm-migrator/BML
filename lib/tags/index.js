const base = require('./base');
const primitives = require('./primitives');
const strings = require('./strings');
const enums = require('./enums');
const array = require('./array');
const sequence = require('./sequence');
const struct = require('./struct');
const constant = require('./constant');
const tap = require('./tap');
const when = require('./when');

module.exports = {
  ...base,
  ...primitives,
  ...strings,
  ...enums,
  ...array,
  ...sequence,
  ...struct,
  ...constant,
  ...tap,
  ...when,
};
