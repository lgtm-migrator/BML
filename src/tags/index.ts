export * from './tag';
export * from './adapter';
export * from './primitives';
export * from './strings';
export * from './enums';
export * from './array';
export * from './sequence';
export * from './struct';
export * from './constant';
export * from './tap';
export * from './when';
export * from './pointer';
export * from './seek';
export * from './skip';
export * from './pass';
export * from './branch';
export * from './computed';
export * from './repeat';
export * from './bitMask';
export * from './endian';
export * from './flag';
export * from './peek';
