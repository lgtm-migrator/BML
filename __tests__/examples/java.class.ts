import * as path from 'path';
import { classFile } from '../../examples/java.class.bml';
import { testFile } from '../helpers/tags';

describe('java class files: ', () => {
  it('should parse', () => {
    testFile(path.join(__dirname, 'files/Test.class'), classFile, 'Test class');
    testFile(
      path.join(__dirname, 'files/Person.class'),
      classFile,
      'Person interface'
    );
    testFile(
      path.join(__dirname, 'files/Student.class'),
      classFile,
      'Student class'
    );
    testFile(
      path.join(__dirname, 'files/Student$SubClass.class'),
      classFile,
      'Student sub class'
    );
  });
});
