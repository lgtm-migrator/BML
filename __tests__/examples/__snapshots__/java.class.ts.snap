// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`java class files:  should parse: Person interface 1`] = `
Object {
  "access_flags": Object {
    "ACC_ABSTRACT": true,
    "ACC_ANNOTATION": false,
    "ACC_ENUM": false,
    "ACC_FINAL": false,
    "ACC_INTERFACE": true,
    "ACC_MODULE": false,
    "ACC_PUBLIC": false,
    "ACC_SUPER": false,
    "ACC_SYNTHETIC": false,
  },
  "attributes": Array [
    Object {
      "attribute_length": 2,
      "attribute_name_index": 5,
      "info": Object {
        "sourcefile_index": 6,
      },
      "name": "SourceFile",
    },
  ],
  "constantpool": Array [
    Object {
      "info": Object {
        "name_index": 7,
      },
      "tag": "constantClass",
    },
    Object {
      "info": Object {
        "name_index": 8,
      },
      "tag": "constantClass",
    },
    Object {
      "info": Object {
        "string": "identify",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "(Ljava/lang/String;)Ljava/lang/String;",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "SourceFile",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "Test.java",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "Person",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "java/lang/Object",
      },
      "tag": "constantUtf8",
    },
  ],
  "constantpool_count": 9,
  "fields": Array [],
  "interfaces": Array [],
  "magic": 3405691582,
  "major_version": 52,
  "methods": Array [
    Object {
      "access_flags": Object {
        "ACC_ABSTRACT": true,
        "ACC_BRIDGE": false,
        "ACC_FINAL": false,
        "ACC_NATIVE": false,
        "ACC_PRIVATE": false,
        "ACC_PROTECTED": false,
        "ACC_PUBLIC": true,
        "ACC_STATIC": false,
        "ACC_STRICT": false,
        "ACC_SYNCHRONIZED": false,
        "ACC_SYNTHETIC": false,
        "ACC_VARARGS": false,
      },
      "attributes": Array [],
      "descriptor_index": 4,
      "name_index": 3,
    },
  ],
  "minor_version": 0,
  "super_class": 2,
  "this_class": 1,
}
`;

exports[`java class files:  should parse: Student class 1`] = `
Object {
  "access_flags": Object {
    "ACC_ABSTRACT": false,
    "ACC_ANNOTATION": false,
    "ACC_ENUM": false,
    "ACC_FINAL": false,
    "ACC_INTERFACE": false,
    "ACC_MODULE": false,
    "ACC_PUBLIC": false,
    "ACC_SUPER": true,
    "ACC_SYNTHETIC": false,
  },
  "attributes": Array [
    Object {
      "attribute_length": 2,
      "attribute_name_index": 14,
      "info": Object {
        "sourcefile_index": 15,
      },
      "name": "SourceFile",
    },
    Object {
      "attribute_length": 10,
      "attribute_name_index": 7,
      "info": Object {
        "classes": Array [
          Object {
            "inner_class_access_flags": Object {
              "ACC_ABSTRACT": false,
              "ACC_ANNOTATION": false,
              "ACC_ENUM": false,
              "ACC_FINAL": false,
              "ACC_INTERFACE": false,
              "ACC_PRIVATE": false,
              "ACC_PROTECTED": false,
              "ACC_PUBLIC": false,
              "ACC_STATIC": false,
              "ACC_SYNTHETIC": false,
            },
            "inner_class_info_index": 5,
            "inner_name_index": 6,
            "outer_class_info_index": 2,
          },
        ],
      },
      "name": "InnerClasses",
    },
  ],
  "constantpool": Array [
    Object {
      "info": Object {
        "class_index": 3,
        "name_and_type_index": 16,
      },
      "tag": "constantMethodref",
    },
    Object {
      "info": Object {
        "name_index": 17,
      },
      "tag": "constantClass",
    },
    Object {
      "info": Object {
        "name_index": 18,
      },
      "tag": "constantClass",
    },
    Object {
      "info": Object {
        "name_index": 19,
      },
      "tag": "constantClass",
    },
    Object {
      "info": Object {
        "name_index": 20,
      },
      "tag": "constantClass",
    },
    Object {
      "info": Object {
        "string": "SubClass",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "InnerClasses",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "<init>",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "()V",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "Code",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "LineNumberTable",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "identify",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "(Ljava/lang/String;)Ljava/lang/String;",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "SourceFile",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "Test.java",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "descriptor_index": 9,
        "name_index": 8,
      },
      "tag": "constantNameAndType",
    },
    Object {
      "info": Object {
        "string": "Student",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "java/lang/Object",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "Person",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "Student$SubClass",
      },
      "tag": "constantUtf8",
    },
  ],
  "constantpool_count": 21,
  "fields": Array [],
  "interfaces": Array [
    4,
  ],
  "magic": 3405691582,
  "major_version": 52,
  "methods": Array [
    Object {
      "access_flags": Object {
        "ACC_ABSTRACT": false,
        "ACC_BRIDGE": false,
        "ACC_FINAL": false,
        "ACC_NATIVE": false,
        "ACC_PRIVATE": true,
        "ACC_PROTECTED": false,
        "ACC_PUBLIC": false,
        "ACC_STATIC": false,
        "ACC_STRICT": false,
        "ACC_SYNCHRONIZED": false,
        "ACC_SYNTHETIC": false,
        "ACC_VARARGS": false,
      },
      "attributes": Array [
        Object {
          "attribute_length": 33,
          "attribute_name_index": 10,
          "info": Object {
            "attributes": Array [
              Object {
                "attribute_length": 10,
                "attribute_name_index": 11,
                "info": Object {
                  "count": 2,
                  "line_number_table": Array [
                    Object {
                      "line_number": 9,
                      "start_pc": 0,
                    },
                    Object {
                      "line_number": 10,
                      "start_pc": 4,
                    },
                  ],
                },
                "name": "LineNumberTable",
              },
            ],
            "code": Array [
              42,
              183,
              0,
              1,
              177,
            ],
            "exception_table": Array [],
            "max_locals": 1,
            "max_stack": 1,
          },
          "name": "Code",
        },
      ],
      "descriptor_index": 9,
      "name_index": 8,
    },
    Object {
      "access_flags": Object {
        "ACC_ABSTRACT": false,
        "ACC_BRIDGE": false,
        "ACC_FINAL": false,
        "ACC_NATIVE": false,
        "ACC_PRIVATE": false,
        "ACC_PROTECTED": false,
        "ACC_PUBLIC": true,
        "ACC_STATIC": false,
        "ACC_STRICT": false,
        "ACC_SYNCHRONIZED": false,
        "ACC_SYNTHETIC": false,
        "ACC_VARARGS": false,
      },
      "attributes": Array [
        Object {
          "attribute_length": 26,
          "attribute_name_index": 10,
          "info": Object {
            "attributes": Array [
              Object {
                "attribute_length": 6,
                "attribute_name_index": 11,
                "info": Object {
                  "count": 1,
                  "line_number_table": Array [
                    Object {
                      "line_number": 13,
                      "start_pc": 0,
                    },
                  ],
                },
                "name": "LineNumberTable",
              },
            ],
            "code": Array [
              43,
              176,
            ],
            "exception_table": Array [],
            "max_locals": 2,
            "max_stack": 1,
          },
          "name": "Code",
        },
      ],
      "descriptor_index": 13,
      "name_index": 12,
    },
  ],
  "minor_version": 0,
  "super_class": 3,
  "this_class": 2,
}
`;

exports[`java class files:  should parse: Student sub class 1`] = `
Object {
  "access_flags": Object {
    "ACC_ABSTRACT": false,
    "ACC_ANNOTATION": false,
    "ACC_ENUM": false,
    "ACC_FINAL": false,
    "ACC_INTERFACE": false,
    "ACC_MODULE": false,
    "ACC_PUBLIC": false,
    "ACC_SUPER": true,
    "ACC_SYNTHETIC": false,
  },
  "attributes": Array [
    Object {
      "attribute_length": 2,
      "attribute_name_index": 11,
      "info": Object {
        "sourcefile_index": 12,
      },
      "name": "SourceFile",
    },
    Object {
      "attribute_length": 10,
      "attribute_name_index": 18,
      "info": Object {
        "classes": Array [
          Object {
            "inner_class_access_flags": Object {
              "ACC_ABSTRACT": false,
              "ACC_ANNOTATION": false,
              "ACC_ENUM": false,
              "ACC_FINAL": false,
              "ACC_INTERFACE": false,
              "ACC_PRIVATE": false,
              "ACC_PROTECTED": false,
              "ACC_PUBLIC": false,
              "ACC_STATIC": false,
              "ACC_SYNTHETIC": false,
            },
            "inner_class_info_index": 3,
            "inner_name_index": 17,
            "outer_class_info_index": 15,
          },
        ],
      },
      "name": "InnerClasses",
    },
  ],
  "constantpool": Array [
    Object {
      "info": Object {
        "class_index": 3,
        "name_and_type_index": 13,
      },
      "tag": "constantFieldref",
    },
    Object {
      "info": Object {
        "class_index": 4,
        "name_and_type_index": 14,
      },
      "tag": "constantMethodref",
    },
    Object {
      "info": Object {
        "name_index": 16,
      },
      "tag": "constantClass",
    },
    Object {
      "info": Object {
        "name_index": 19,
      },
      "tag": "constantClass",
    },
    Object {
      "info": Object {
        "string": "this$0",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "LStudent;",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "<init>",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "(LStudent;)V",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "Code",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "LineNumberTable",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "SourceFile",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "Test.java",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "descriptor_index": 6,
        "name_index": 5,
      },
      "tag": "constantNameAndType",
    },
    Object {
      "info": Object {
        "descriptor_index": 20,
        "name_index": 7,
      },
      "tag": "constantNameAndType",
    },
    Object {
      "info": Object {
        "name_index": 21,
      },
      "tag": "constantClass",
    },
    Object {
      "info": Object {
        "string": "Student$SubClass",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "SubClass",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "InnerClasses",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "java/lang/Object",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "()V",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "Student",
      },
      "tag": "constantUtf8",
    },
  ],
  "constantpool_count": 22,
  "fields": Array [
    Object {
      "access_flags": Object {
        "ACC_ENUM": false,
        "ACC_FINAL": true,
        "ACC_PRIVATE": false,
        "ACC_PROTECTED": false,
        "ACC_PUBLIC": false,
        "ACC_STATIC": false,
        "ACC_SYNTHETIC": true,
        "ACC_TRANSIENT": false,
        "ACC_VOLATILE": false,
      },
      "attributes": Array [],
      "descriptor_index": 6,
      "name_index": 5,
    },
  ],
  "interfaces": Array [],
  "magic": 3405691582,
  "major_version": 52,
  "methods": Array [
    Object {
      "access_flags": Object {
        "ACC_ABSTRACT": false,
        "ACC_BRIDGE": false,
        "ACC_FINAL": false,
        "ACC_NATIVE": false,
        "ACC_PRIVATE": false,
        "ACC_PROTECTED": false,
        "ACC_PUBLIC": false,
        "ACC_STATIC": false,
        "ACC_STRICT": false,
        "ACC_SYNCHRONIZED": false,
        "ACC_SYNTHETIC": false,
        "ACC_VARARGS": false,
      },
      "attributes": Array [
        Object {
          "attribute_length": 34,
          "attribute_name_index": 9,
          "info": Object {
            "attributes": Array [
              Object {
                "attribute_length": 6,
                "attribute_name_index": 10,
                "info": Object {
                  "count": 1,
                  "line_number_table": Array [
                    Object {
                      "line_number": 6,
                      "start_pc": 0,
                    },
                  ],
                },
                "name": "LineNumberTable",
              },
            ],
            "code": Array [
              42,
              43,
              181,
              0,
              1,
              42,
              183,
              0,
              2,
              177,
            ],
            "exception_table": Array [],
            "max_locals": 2,
            "max_stack": 2,
          },
          "name": "Code",
        },
      ],
      "descriptor_index": 8,
      "name_index": 7,
    },
  ],
  "minor_version": 0,
  "super_class": 4,
  "this_class": 3,
}
`;

exports[`java class files:  should parse: Test class 1`] = `
Object {
  "access_flags": Object {
    "ACC_ABSTRACT": false,
    "ACC_ANNOTATION": false,
    "ACC_ENUM": false,
    "ACC_FINAL": false,
    "ACC_INTERFACE": false,
    "ACC_MODULE": false,
    "ACC_PUBLIC": false,
    "ACC_SUPER": true,
    "ACC_SYNTHETIC": false,
  },
  "attributes": Array [
    Object {
      "attribute_length": 2,
      "attribute_name_index": 13,
      "info": Object {
        "sourcefile_index": 14,
      },
      "name": "SourceFile",
    },
  ],
  "constantpool": Array [
    Object {
      "info": Object {
        "class_index": 6,
        "name_and_type_index": 15,
      },
      "tag": "constantMethodref",
    },
    Object {
      "info": Object {
        "class_index": 16,
        "name_and_type_index": 17,
      },
      "tag": "constantFieldref",
    },
    Object {
      "info": Object {
        "string_index": 18,
      },
      "tag": "constantString",
    },
    Object {
      "info": Object {
        "class_index": 19,
        "name_and_type_index": 20,
      },
      "tag": "constantMethodref",
    },
    Object {
      "info": Object {
        "name_index": 21,
      },
      "tag": "constantClass",
    },
    Object {
      "info": Object {
        "name_index": 22,
      },
      "tag": "constantClass",
    },
    Object {
      "info": Object {
        "string": "<init>",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "()V",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "Code",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "LineNumberTable",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "main",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "([Ljava/lang/String;)V",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "SourceFile",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "Test.java",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "descriptor_index": 8,
        "name_index": 7,
      },
      "tag": "constantNameAndType",
    },
    Object {
      "info": Object {
        "name_index": 23,
      },
      "tag": "constantClass",
    },
    Object {
      "info": Object {
        "descriptor_index": 25,
        "name_index": 24,
      },
      "tag": "constantNameAndType",
    },
    Object {
      "info": Object {
        "string": "Class File Structure",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "name_index": 26,
      },
      "tag": "constantClass",
    },
    Object {
      "info": Object {
        "descriptor_index": 28,
        "name_index": 27,
      },
      "tag": "constantNameAndType",
    },
    Object {
      "info": Object {
        "string": "Test",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "java/lang/Object",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "java/lang/System",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "out",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "Ljava/io/PrintStream;",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "java/io/PrintStream",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "println",
      },
      "tag": "constantUtf8",
    },
    Object {
      "info": Object {
        "string": "(Ljava/lang/String;)V",
      },
      "tag": "constantUtf8",
    },
  ],
  "constantpool_count": 29,
  "fields": Array [],
  "interfaces": Array [],
  "magic": 3405691582,
  "major_version": 52,
  "methods": Array [
    Object {
      "access_flags": Object {
        "ACC_ABSTRACT": false,
        "ACC_BRIDGE": false,
        "ACC_FINAL": false,
        "ACC_NATIVE": false,
        "ACC_PRIVATE": false,
        "ACC_PROTECTED": false,
        "ACC_PUBLIC": false,
        "ACC_STATIC": false,
        "ACC_STRICT": false,
        "ACC_SYNCHRONIZED": false,
        "ACC_SYNTHETIC": false,
        "ACC_VARARGS": false,
      },
      "attributes": Array [
        Object {
          "attribute_length": 29,
          "attribute_name_index": 9,
          "info": Object {
            "attributes": Array [
              Object {
                "attribute_length": 6,
                "attribute_name_index": 10,
                "info": Object {
                  "count": 1,
                  "line_number_table": Array [
                    Object {
                      "line_number": 18,
                      "start_pc": 0,
                    },
                  ],
                },
                "name": "LineNumberTable",
              },
            ],
            "code": Array [
              42,
              183,
              0,
              1,
              177,
            ],
            "exception_table": Array [],
            "max_locals": 1,
            "max_stack": 1,
          },
          "name": "Code",
        },
      ],
      "descriptor_index": 8,
      "name_index": 7,
    },
    Object {
      "access_flags": Object {
        "ACC_ABSTRACT": false,
        "ACC_BRIDGE": false,
        "ACC_FINAL": false,
        "ACC_NATIVE": false,
        "ACC_PRIVATE": false,
        "ACC_PROTECTED": false,
        "ACC_PUBLIC": true,
        "ACC_STATIC": true,
        "ACC_STRICT": false,
        "ACC_SYNCHRONIZED": false,
        "ACC_SYNTHETIC": false,
        "ACC_VARARGS": false,
      },
      "attributes": Array [
        Object {
          "attribute_length": 37,
          "attribute_name_index": 9,
          "info": Object {
            "attributes": Array [
              Object {
                "attribute_length": 10,
                "attribute_name_index": 10,
                "info": Object {
                  "count": 2,
                  "line_number_table": Array [
                    Object {
                      "line_number": 20,
                      "start_pc": 0,
                    },
                    Object {
                      "line_number": 21,
                      "start_pc": 8,
                    },
                  ],
                },
                "name": "LineNumberTable",
              },
            ],
            "code": Array [
              178,
              0,
              2,
              18,
              3,
              182,
              0,
              4,
              177,
            ],
            "exception_table": Array [],
            "max_locals": 1,
            "max_stack": 2,
          },
          "name": "Code",
        },
      ],
      "descriptor_index": 12,
      "name_index": 11,
    },
  ],
  "minor_version": 0,
  "super_class": 6,
  "this_class": 5,
}
`;
